openhands/runtime/utils/bash.py

```diff
@@ -66,269 +68,402 @@ def split_bash_commands(commands):
     return result
 
 
+class BashCommandStatus(Enum):
+    CONTINUE = 'continue'
+    COMPLETED = 'completed'
+    NO_CHANGE_TIMEOUT = 'no_change_timeout'
+    HARD_TIMEOUT = 'hard_timeout'
+
+
+def _remove_command_prefix(command_output: str, command: str) -> str:
+    return command_output.lstrip().removeprefix(command.lstrip()).lstrip()
+
+
 class BashSession:
-    """A class that maintains a pexpect process and provides a simple interface for running commands and interacting with the shell."""
+    POLL_INTERVAL = 0.5
+    HISTORY_LIMIT = 10_000
+    PS1 = CmdOutputMetadata.to_ps1_prompt()
 
-    def __init__(self, work_dir: str, username: str):
-        self._pwd = work_dir
+    def __init__(
+        self,
+        work_dir: str,
+        username: str | None = None,
+        no_change_timeout_seconds: float = 30.0,
+    ):
+        self.NO_CHANGE_TIMEOUT_SECONDS = no_change_timeout_seconds
+
+        self.server = libtmux.Server()
+        window_command = '/bin/bash'
+        if username:
+            # This starts a non-login (new) shell for the given user
+            window_command = f'su {username} -'
+
+        session_name = f'openhands-{username}-{uuid.uuid4()}'
+        self.session = self.server.new_session(
```

Comment:
how much work is this doing? We should probably move this into a `connect()` method instead of doing it in __init__

Reply:
I believe this piece of code is running inside the runtime docker container. I didn't measure the performance but I think it wasn't *that* bad?

Reply:
done in most recent commit -- move it to a separate .initialize function and now initialize in `ainit()` of action execution server